crates/ark/src/interface.rs

```diff
@@ -585,33 +589,26 @@ impl RMain {
 
     /// Access a reference to the singleton instance of this struct
     ///
-    /// SAFETY: Accesses must occur after `RMain::start()` initializes it, and must
-    /// occur on the main R thread.
+    /// SAFETY: Accesses must occur after `RMain::start()` initializes it.
     pub fn get() -> &'static Self {
         RMain::get_mut()
     }
 
     /// Access a mutable reference to the singleton instance of this struct
     ///
-    /// SAFETY: Accesses must occur after `RMain::start()` initializes it, and must
-    /// occur on the main R thread.
+    /// SAFETY: Accesses must occur after `RMain::start()` initializes it.
+    /// Be aware that we're bypassing the borrow checker. The only guarantee we
+    /// have is that `R_MAIN` is only accessed from the R thread. If you're
+    /// inspecting mutable state, or mutating state, you must reason the
+    /// soundness by yourself.
     pub fn get_mut() -> &'static mut Self {
-        if !RMain::on_main_thread() {
-            let thread = std::thread::current();
-            let name = thread.name().unwrap_or("<unnamed>");
-            let message =
-                format!("Must access `R_MAIN` on the main R thread, not thread '{name}'.");
-            #[cfg(debug_assertions)]
-            panic!("{message}");
-            #[cfg(not(debug_assertions))]
-            log::error!("{message}");
-        }
+        R_MAIN.with_borrow_mut(|cell| {
+            let main_ref = cell.get_mut();
 
-        unsafe {
-            R_MAIN
-                .as_mut()
-                .expect("`R_MAIN` can't be used before it is initialized!")
-        }
+            // We extend the lifetime to `'static` as `R_MAIN` is effectively static once initialized.
+            // This allows us to return a `&mut` from the unsafe cell to the caller.
+            unsafe { std::mem::transmute::<&mut RMain, &'static mut RMain>(main_ref) }
```

Comment:
The alternative is to remove `get_mut()` in favor of the `with_mut()` everywhere right?

Since effectively what we are bypassing here is "the reference may outlive the object" kind of compiler errors?

Reply:
Correct. I went for minimal changes here.

Reply:
Also I don't think this transmute is problematic since `R_MAIN` is effectively `'static` after init.